### DISCUSSION TITLE
The `assert()` method's behavior is different between 3.x and 4.0 .

### DISCUSSION THREAD

**Discussion by Daylily-Zeleen** [godotengine/godot-proposals#5455] [Engine Core]
In 4.0, `assert()` is second argument `message` must be a static `String`, but in 3.x, it can be dynamic.

For example:
```
var list: Array = []

func get_value(idx: int):
    # Valid in 3.x, invalid in 4.0
    assert(idx < list.size() and idx > 0, "Invalid index: %d"%idx)
    return list[idx]
```

Of course, we can use other way to express the detail error message for the `assert()`, like code a comment near the `assert()`,
but obviously, implement dynamic `String` for `message` is the best way.

What is the reson why 4.0 let `message` static?

**Comment by Zireael07** [comment]
is a bug: https://github.com/godotengine/godot/issues/47157

**Reply by Daylily-Zeleen** [reply]
But it is written in help document.

![image](https://user-images.githubusercontent.com/61624558/190980883-0b681d8d-7868-4f49-8dfa-672987bbb9df.png)

**Reply by Calinou** [reply]
> But it is written in help document.

A documented bug can still be a bug :slightly_smiling_face: 

I think someone documented this behavior thinking it was intentional, but it remains a bug nonetheless.

**Reply by Daylily-Zeleen** [reply]
I see. Thanks for your reply.
